pkg/daemon/daemon.go

```diff
@@ -317,72 +319,92 @@ func cmdRun(p *ptpProcess, stdoutToSocket bool) {
-		return
-	}
-
 	done := make(chan struct{})
-
-	if !stdoutToSocket {
-		scanner := bufio.NewScanner(cmdReader)
-		go func() {
-			for scanner.Scan() {
-				output := scanner.Text()
-				fmt.Printf("%s\n", output)
-				extractMetrics(p.configName, p.name, p.ifaces, output)
-			}
-			done <- struct{}{}
-		}()
-	} else {
-		go func() {
-		connect:
-			select {
-			case <-p.exitCh:
+	for {
+		glog.Infof("Starting %s...", p.name)
+		glog.Infof("%s cmd: %+v", p.name, p.cmd)
+
+		//
+		// don't discard process stderr output
+		//
+		p.cmd.Stderr = os.Stderr
+		cmdReader, err := p.cmd.StdoutPipe()
+		if err != nil {
+			glog.Errorf("cmdRun() error creating StdoutPipe for %s: %v", p.name, err)
+			return
+		}
+		if !stdoutToSocket {
+			scanner := bufio.NewScanner(cmdReader)
+			go func() {
+				for scanner.Scan() {
+					output := scanner.Text()
+					fmt.Printf("%s\n", output)
+					extractMetrics(p.configName, p.name, p.ifaces, output)
+				}
 				done <- struct{}{}
-			default:
-				c, err = net.Dial("unix", eventSocket)
-				if err != nil {
-					glog.Errorf("error trying to connect to event socket")
-					time.Sleep(connectionRetryInterval)
-					goto connect
+			}()
+		} else {
+			go func() {
+			connect:
+				select {
+				case <-p.exitCh:
+					done <- struct{}{}
+				default:
+					c, err = net.Dial("unix", eventSocket)
+					if err != nil {
+						glog.Errorf("error trying to connect to event socket")
+						time.Sleep(connectionRetryInterval)
+						goto connect
+					}
 				}
-			}
-			scanner := bufio.NewScanner(cmdReader)
-			for scanner.Scan() {
-				output := scanner.Text()
-				out := fmt.Sprintf("%s\n", output)
-				fmt.Printf("%s", out)
-				_, err := c.Write([]byte(out))
-				if err != nil {
-					glog.Errorf("Write error:", err)
-					goto connect
+				scanner := bufio.NewScanner(cmdReader)
+				for scanner.Scan() {
+					output := scanner.Text()
+					out := fmt.Sprintf("%s\n", output)
+					fmt.Printf("%s", out)
+					_, err := c.Write([]byte(out))
+					if err != nil {
+						glog.Errorf("Write error:", err)
+						goto connect
+					}
 				}
+				done <- struct{}{}
+			}()
+		}
+		p.restartMutex.Lock() // Don't restart after termination
+		restart := p.restart
+		if restart {
+			err = p.cmd.Start()
+			p.restartMutex.Unlock()
+			if err != nil {
+				glog.Errorf("cmdRun() error starting %s: %v", p.name, err)
+				goto recreateCmd
 			}
-			done <- struct{}{}
-		}()
-	}
-
-	err = p.cmd.Start()
-	if err != nil {
-		glog.Errorf("cmdRun() error starting %s: %v", p.name, err)
-		return
-	}
-
-	<-done
+		} else {
+			p.restartMutex.Unlock()
+		}
 
-	err = p.cmd.Wait()
-	if err != nil {
-		glog.Errorf("cmdRun() error waiting for %s: %v", p.name, err)
-		return
+		<-done
+		if restart {
+			err = p.cmd.Wait()
+			if err != nil {
+				glog.Errorf("cmdRun() error waiting for %s: %v", p.name, err)
+			}
+		}
+	recreateCmd: // Recreate command after it terminates.
```

Comment:
can remove recreateCmd lable

Reply:
Was there to skip waiting if error.  Switching to a bool

Reply:
Previously it was returning on error since now you don't want  to return here and check for recreating if you don't need that return and don't need goto label